app/controllers/application_controller.rb

```diff
@@ -12,6 +12,8 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
 
+  around_filter :profile, if: -> {Rails.env.development? && params[:trace] == "1"}
```

Comment:
Before filters run code before running the current route's action. After filters fun code after the current route's action. Around filters are a bit trickier in that they run it *around* the action, meaning both before and after.

The basic idea here is that before it starts running the action, it runs ```profile``` up to the point where the profile method calls ```yield```, at which point it runs the action. Once the action is done, it returns to profile and keeps running from the point ```yield``` was called.

So, in this case, it runs

```ruby
require 'ruby-prof'

RubyProf.start

yield
```

This starts the profile recording. Then it runs the current action for the route. And then after the action is finished, it runs the rest of the method, which stops the profiler's recording, and prints out a graph of the results

```ruby
results = RubyProf.stop

File.open "#{Rails.root}/profile-stack.html", 'w' do |file|
  RubyProf::CallStackPrinter.new(results).print(file)
end
```

There's more explanation in [the Rails Guide for ActionController](http://guides.rubyonrails.org/action_controller_overview.html#after-filters-and-around-filters).